### PR TITLE
Add support for initializers declared as throws

### PR DESCRIPTION
**Pull request by lydia-duncan**
This is a very early start at supporting throwing from initializers.  It has a number of
restrictions on it as a result, but is hopefully sufficient to unblock some 2.0 tasks
that were best resolved by throwing initializers.

The list of what is supported by this PR:
- Adding the `throws` keyword to the initializer declaration
- Calling functions that throw during Phase 2 of the initializer body

Some things that are not yet supported:
- Calling functions that throw during Phase 0 or 1 of the initializer body
  - Note that this includes some functionality that is not as obvious, such as
loops or some casts.  This limitation is only on initializers declared with the
keyword
- Explicitly throwing an error in the body
- Using a try or try! with a catch block

Things that were supported before and still are:
- Using `try!` without a catch block anywhere in the body
- Calling functions that throw at any point in initializers that are declared
without the `throws` keyword in prototype modules

It also adds a missing check to ensure `postinit` cannot be declared as
`throws`, which was allowed but would cause a halt prior to this PR -
the implementation in this PR would cause its behavior to be inconsistent
depending on if there was also an initializer that throws, so just blanket
ban it until we can do a proper implementation.

The rules for initializers are similarly applied to copy initializers since the
detection is mostly the same.

Implementation details:

<details>

- Adds a primitive, PRIM_INIT_DONE, so that calls to `this.complete()`
in the body of the initializer are still maintained and recognized during
error handling
- Adds a new state tracker for use during error handling that will track
which Phase an initializer is in, updating appropriately for conditionals
and recognizing `this.init` and `super.init` calls at that stage in
compilation.
- Uses this state tracker in the error handling pass, which required
adding `enterCondStmt` to the error handling visitor
- Removed error during normalize for declaring initializers as throws
- Adds a check to ensure `postinit` can't be declared as throws
- Adjust the creation of the `_new` function when the initializer throws
to catch the error, clean up the type, and then rethrow the error, allowing
us to appropriately manage the memory.

</details>

Testing:

Updates the test of declaring an initializer as throws but not actually throwing
from it now that it is supported

Also adds a number of tests of this feature:
- Calling functions that throw during Phase 1 (not allowed)
- Calling functions that throw during Phase 0 (not allowed)
- Calling functions that throw from both branches of a conditional during Phase 2
- Calling functions that throw from the then branch of a conditional during
Phase 1 (not allowed)
- Calling functions that throw from the else branch of a conditional during Phase 1
(not allowed)
- Inheriting from a class with a throwing initializer and trying to call that
initializer using `super.init()` (not allowed)
- Calling functions that throw and letting that throw to main
- Calling functions that throw and letting that throw to a function that called the
throwing initializer
- A field default value that is made by calling a throwing function
- copy initializers declared as throws that don't actually throw
- copy initializers that throw out to main in the same way as the initializer case above
- copy initializers that throw out to a function that called them (ditto previous bullet)
- copy initializers that attempt to calling throwing functions during Phase 1 (not
allowed)
- A future to track declaring deinitializers as `throws`
- A future to track declaring postinit as `throws`

Passed a full paratest with futures

### CODE REVIEW

**Comment by lydia-duncan** [comment]
I see the copyright issue, should be an easy fix

**Review by mppf — APPROVED** [review]
I looked over the compiler changes but not the tests & things look good to me.

File: compiler/passes/InitErrorHandling.cpp [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+*                                                                             *
+*                                                                             *
+*                                                                             *
+************************************** | *************************************/
+static const char* initName(CallExpr* stmt);
+
+bool isResolvedSuperInit(CallExpr* stmt) {
+  const char* name = initName(stmt);
+
+  return name != NULL && strcmp(name, "super") == 0 ? true : false;
```
Comment (mppf) [inline compiler/passes/InitErrorHandling.cpp:81]: can you use `call->isNamedAstr(astrSuper)` ?

Comment (mppf) [reply compiler/passes/InitErrorHandling.cpp:81]: Or maybe just `name == astrSuper`?

File: compiler/passes/InitErrorHandling.cpp [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+
+static const char* initName(CallExpr* stmt) {
+  const char* retval = NULL;
+
+  // Not in a primitive
+  if (stmt->baseExpr != NULL) {
+    if (SymExpr* base = toSymExpr(stmt->baseExpr)) {
+
+      // Call is to an initializer
+      if (strcmp(base->symbol()->name, "init") == 0 &&
```
Comment (mppf) [inline compiler/passes/InitErrorHandling.cpp:106]: `base->symbol()->name = astrInit`

File: compiler/passes/InitErrorHandling.cpp [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+          stmt->numActuals() >= 1) {
+        SymExpr* firstArg = toSymExpr(stmt->get(1));
+        if (strcmp(firstArg->symbol()->name, "super") == 0 ||
```
Comment (mppf) [inline compiler/passes/InitErrorHandling.cpp:109]: these strcmps seem a bit hacky. Could you put a comment to indicate they are a workaround for something-or-other?

File: compiler/passes/InitErrorHandling.cpp [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+            strcmp(firstArg->symbol()->name, "super_tmp") == 0) {
+          retval = "super";
+        } else if (firstArg->symbol()->hasFlag(FLAG_ARG_THIS)) {
+          retval = "this";
+        }
+      }
+    } else if (UnresolvedSymExpr* urse = toUnresolvedSymExpr(stmt->baseExpr)) {
+      if (strcmp(urse->unresolved, "init") == 0) {
```
Comment (mppf) [inline compiler/passes/InitErrorHandling.cpp:117]: `urse->unresolved == astrInit`

**Pull request merged**